src/main/java/io/vertx/core/net/impl/InboundMessageQueue.java

```diff
@@ -0,0 +1,206 @@
+      readQueueFactory = InboundReadQueue.SINGLE_THREADED;
+    } else {
+      readQueueFactory = InboundReadQueue.SPSC;
+    }
+    this.readQueue = readQueueFactory.create(this, lowWaterMark, highWaterMark);
+    this.context = context;
+    this.eventLoop = eventLoop;
+  }
+
+  @Override
+  public final boolean test(M msg) {
+    while (true) {
+      long d = DEMAND_UPDATER.get(this);
+      if (d == 0L) {
+        return false;
+      } else if (d == Long.MAX_VALUE || DEMAND_UPDATER.compareAndSet(this, d, d - 1)) {
+        break;
+      }
+    }
+    handleMessage(msg);
+    return true;
+  }
+
+  /**
+   * Handle resume, executed on the event-loop thread.
+   */
+  protected void handleResume() {
+  }
+
+  /**
+   * Handler pause, executed on the event-loop thread
+   */
+  protected void handlePause() {
+  }
+
+  /**
+   * Handle a message, executed on the context thread
+   *
+   * @param msg the message
+   */
+  protected void handleMessage(M msg) {
+  }
+
+  /**
+   * Add a message to the queue
+   *
+   * @param msg the message
+   * @return {@code true} when a {@link #drain()} should be called.
+   */
+  public final boolean add(M msg) {
+    assert eventLoop.inEventLoop();
+    int res = readQueue.add(msg);
+    if ((res & InboundReadQueue.QUEUE_UNWRITABLE_MASK) != 0) {
+      handlePause();
+    }
+    return (res & InboundReadQueue.DRAIN_REQUIRED_MASK) != 0;
+  }
+
+  /**
+   * {@link #add(Object)} + {@link #drain()}.
+   *
+   * @param messages the messages
+   */
+  public final void write(Iterable<M> messages) {
+    boolean drain = false;
+    for (M msg : messages) {
+      drain |= add(msg);
+    }
+    if (drain) {
+      drain();
+    }
+  }
+
+  /**
+   * {@link #add(Object)} + {@link #drain()}.
+   *
+   * @param msg the message
+   */
+  public final void write(M msg) {
+    if (add(msg)) {
+      drain();
+    }
+  }
+
+  /**
+   * Schedule a drain operation on the context thread.
+   */
+  public final void drain() {
+    assert eventLoop.inEventLoop();
+    if (context.inThread()) {
+      drainInternal();
+    } else {
+      context.execute(this::drainInternal);
+    }
+  }
+
+  /**
+   * Task executed from context thread.
+   */
+  @Override
+  public void run() {
+    assert context.inThread();
+    if (!draining && needsDrain) {
+      drainInternal();
+    }
+  }
+
+  private void drainInternal() {
+    draining = true;
+    try {
+      int res = readQueue.drain();
+      needsDrain = (res & InboundReadQueue.DRAIN_REQUIRED_MASK) != 0;
+      if ((res & InboundReadQueue.QUEUE_WRITABLE_MASK) != 0) {
+        eventLoop.execute(this::handleResume);
+      }
+    } finally {
+      draining = false;
+    }
+  }
+
+  /**
+   * Clear the demand.
+   */
+  public final void pause() {
+    DEMAND_UPDATER.set(this, 0L);
+  }
+
+  /**
+   * Add {@code amount} to the current demand.
+   *
+   * @param amount the number of message to consume
+   */
+  public final void fetch(long amount) {
+    if (amount < 0L) {
```

Comment:
You might want to handle `0` as well and reject it.

Reply:
I think it's fine to tolerate 0 is a no-op, any reactive-streams implementation that uses it will never pass 0